File: module2/fiveSort.js
/*
five sort

Write a function, fiveSort, that takes in an array of numbers as an argument. The function should rearrange elements of the array such that all 5s appear at the end. Your function should perform this operation in-place by mutating the original array. The function should return the array.

Elements that are not 5 can appear in any order in the output, as long as all 5s are at the end of the array.
test_00

fiveSort([12, 5, 1, 5, 12, 7]);
// -> [12, 7, 1, 12, 5, 5]

test_01

fiveSort([5, 2, 5, 6, 5, 1, 10, 2, 5, 5]);
// -> [2, 2, 10, 6, 1, 5, 5, 5, 5, 5]
*/

// brute force
const fiveSort = (nums) => {
  let startIdx = 0;
  let endIdx = nums.length - 1;

  while (endIdx > startIdx) {
    if (nums[endIdx] !== 5) {
      if (nums[startIdx] === 5) {
        swap(startIdx, endIdx, nums);
        endIdx -= 1;
        startIdx += 1;
      } else {
        startIdx += 1;
      }
    } else {
      endIdx -= 1;
    }
  }
  return nums;
};

const swap = (i, j, array) => {
  [array[i], array[j]] = [array[j], array[i]];
};

// solution
const fiveSort = (nums) => {
  let startIdx = 0;
  let endIdx = nums.length - 1;

  while (endIdx >= startIdx) {
    if (nums[endIdx] === 5) {
      endIdx -= 1;
    } else if (nums[startIdx] === 5) {
      swap(startIdx, endIdx, nums);
      startIdx += 1;
    } else {
      startIdx += 1;
    }
  }
  return nums;
};

const swap = (i, j, array) => {
  [array[i], array[j]] = [array[j], array[i]];
};

File: module2/five_sort.py
"""
five sort

Write a function, five_sort, that takes in a list of numbers as an argument. The function should rearrange elements of the list such that all 5s appear at the end. Your function should perform this operation in-place by mutating the original list. The function should return the list.

Elements that are not 5 can appear in any order in the output, as long as all 5s are at the end of the list.
test_00

five_sort([12, 5, 1, 5, 12, 7])
# -> [12, 7, 1, 12, 5, 5]

test_01

five_sort([5, 2, 5, 6, 5, 1, 10, 2, 5, 5])
# -> [2, 2, 10, 6, 1, 5, 5, 5, 5, 5]

Approach:
- use 2 pointer strategy with the pointers at the beginning and end of the list
- if end element is 5, decrement
- if start element is not a 5, increment
- swap
"""

# brute force
def five_sort(nums):
  start, end = 0, len(nums) - 1
  while start <= end:
    if nums[end] == 5:
      end -= 1
    elif nums[start] == 5:
        swap(start, end, nums)
        end -= 1
        start += 1
    else:
      start += 1
  return nums


def swap(i, j, array):
  array[i], array[j] = array[j], array[i]

# solution

